File: copilot-demo-labs/use-cases/09-prompt-engineering-core/src/shopping_cart.py
class ShoppingCart:
    def __init__(self):
        self.items = []
        self.discount = 0

    def add(self, name, price, qty):
        for i in self.items:
            if i["name"] == name:
                i["qty"] += qty
                return
        self.items.append({"name": name, "price": price, "qty": qty})

    def remove(self, name):
        self.items = [i for i in self.items if i["name"] != name]

    def set_discount(self, pct):
        self.discount = pct

    def total(self):
        t = 0
        for i in self.items:
            t += i["price"] * i["qty"]
        if self.discount:
            t = t - t * self.discount / 100
        if t < 0:
            t = 0
        return round(t, 2)

    def summary(self):
        lines = []
        for i in self.items:
            lines.append(f"{i['name']} x{i['qty']} = {i['price'] * i['qty']}")
        lines.append(f"Discount: {self.discount}%")
        lines.append(f"Total: {self.total()}")
        return "\n".join(lines)

File: copilot-demo-labs/use-cases/09-prompt-engineering-core/src/ShoppingCart.cs
namespace Demo;

public class CartItem
{
    public string Name { get; set; } = "";
    public decimal Price { get; set; }
    public int Qty { get; set; }
}

public class ShoppingCart
{
    private readonly List<CartItem> _items = new();
    private decimal _discount;

    public void Add(string name, decimal price, int qty)
    {
        var existing = _items.FirstOrDefault(i => i.Name == name);
        if (existing is not null)
        {
            existing.Qty += qty;
            return;
        }
        _items.Add(new CartItem { Name = name, Price = price, Qty = qty });
    }

    public void Remove(string name)
    {
        _items.RemoveAll(i => i.Name == name);
    }

    public void SetDiscount(decimal pct)
    {
        _discount = pct;
    }

    public decimal Total()
    {
        var t = _items.Sum(i => i.Price * i.Qty);
        if (_discount > 0)
        {
            t -= t * _discount / 100;
        }
        if (t < 0) t = 0;
        return Math.Round(t, 2);
    }

    public string Summary()
    {
        var lines = _items.Select(i => $"{i.Name} x{i.Qty} = {i.Price * i.Qty}").ToList();
        lines.Add($"Discount: {_discount}%");
        lines.Add($"Total: {Total()}");
        return string.Join("\n", lines);
    }
}
